File: python_solutions/992.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        l , r1 , r2 = 0 , 0 , 0
        map1 = {}
        arr_cnt = 0
        while l < len(nums):
            while r1 < len(nums) and len(map1) < k:
                if nums[r1] in map1:
                    map1[nums[r1]] += 1
                else:
                    map1[nums[r1]] = 1
                if r1 == r2:
                    r2 += 1
                r1 += 1
            while r2 < len(nums) and nums[r2] in map1:
                r2 += 1
            if len(map1) == k:
                arr_cnt += r2-r1+1
            map1[nums[l]] -= 1
            if map1[nums[l]] == 0:
                map1.pop(nums[l])
            l += 1
            
        return arr_cnt

File: c++_solutions/992.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        unordered_map<int,int> freq ;
        int l = 0, r1 =0 ,r2 = 0 ;
        int arr_cnt = 0;
        while (l < nums.size()){
            while (r1 < nums.size() && freq.size() < k){
                if (freq.find(nums[r1]) != freq.end()){
                    freq[nums[r1]] += 1 ; 
                }
                else{
                    freq[nums[r1]] = 1;
                }
                if (r2 == r1){
                    r2++ ;
                }
                r1++ ;
            }
            while (r2 < nums.size() && freq.find(nums[r2])!=freq.end()){
                r2++;
            }
            if (freq.size() == k){
                arr_cnt += r2-r1+1 ;
            }
            freq[nums[l]] -= 1 ;
            if (freq[nums[l]] == 0){
                freq.erase(nums[l]);
            }
            l++;
        }
        return arr_cnt; 
    }
};
